test/which.js

```diff
@@ -69,5 +70,27 @@ test('Searching with -a flag returns an array with first item equals to the regu
   t.falsy(shell.error());
   t.truthy(resultForWhich);
   t.truthy(resultForWhichA);
-  t.is(resultForWhich.toString(), resultForWhichA[0].toString());
+  t.is(resultForWhich.toString(), resultForWhichA[0]);
+});
+
+test('None executable files does not appear in the result list', t => {
+  const commandName = 'node'; // Should be an existing command
+  const extraPath = path.resolve(__dirname, 'resources', 'which');
+  const matchingFile = path.resolve(extraPath, commandName);
+  const pathEnv = process.env.PATH;
+
+  // make sure that file is exists (will throw error otherwise)
+  t.truthy(fs.existsSync(matchingFile));
+
+  process.env.PATH = extraPath + path.delimiter + process.env.PATH;
+  const resultForWhich = shell.which(commandName);
+  const resultForWhichA = shell.which('-a', commandName);
+  t.falsy(shell.error());
+  t.truthy(resultForWhich);
+  t.truthy(resultForWhichA);
+  t.truthy(resultForWhichA.length);
+  t.not(resultForWhich.toString(), matchingFile);
+  t.is(resultForWhichA.indexOf(matchingFile), -1);
+
+  process.env.PATH = pathEnv;
```

Comment:
Oh sorry, I must have missed this. This should go in an `test.afterEach.always` block. Example:

https://github.com/shelljs/shelljs/blob/6d66a1af98430bc0275e2125d4cae41ccb0de645/test/exec.js#L14-L17

This shouldn't block this change. I added a comment to #834 to track this.